internal/transport/http2_client.go

```diff
@@ -1279,59 +1296,112 @@ func (t *http2Client) reader() {
+			return
+		}
+	}
+}
+
+// pinger sends out a PING frame, and waits for one of the following events to
+// happen (kp.Timeout expires, top-level context is done, or the done channel
+// is written to (this usually happens when there is read activity).
+func (t *http2Client) pinger(send bool, done chan struct{}) {
+	if send {
+		if channelz.IsOn() {
+			atomic.AddInt64(&t.czData.kpCount, 1)
+		}
+		t.controlBuf.put(&ping{data: [8]byte{}})
+	}
+
+	timer := time.NewTimer(t.kp.Timeout)
+	select {
+	case <-timer.C:
+		t.Close()
+		return
+	case <-t.ctx.Done():
+		if !timer.Stop() {
+			<-timer.C
+		}
+		return
+	case <-done:
+		if !timer.Stop() {
+			<-timer.C
+		}
+		return
+	}
+}
+
+// keepalive running in a separate goroutune makes sure the connection is alive
+// by sending pings.
 func (t *http2Client) keepalive() {
-	p := &ping{data: [8]byte{}}
-	timer := time.NewTimer(t.kp.Time)
+	go t.activityMonitor()
+
+	ticker := time.NewTicker(defaultKeepaliveTickerDuration)
+	defer ticker.Stop()
+
+	var done chan struct{}
 	for {
 		select {
-		case <-timer.C:
-			if atomic.CompareAndSwapUint32(&t.activity, 1, 0) {
-				timer.Reset(t.kp.Time)
+		case <-ticker.C:
+			// If the amount of time elapsed since the last read is less than kp.Time
+			// and we have an active pinger routine, we need to cancel it.
+			lr := time.Since(time.Unix(atomic.LoadInt64(&t.lr.time), 0))
+			if lr < t.kp.Time {
+				if done != nil {
+					close(done)
+					done = nil
+				}
 				continue
 			}
-			// Check if keepalive should go dormant.
-			t.mu.Lock()
-			if len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream {
-				// Make awakenKeepalive writable.
-				<-t.awakenKeepalive
-				t.mu.Unlock()
+			// If the control gets here, it means more than kp.Time has elapsed since
+			// the last read, and we need to do one of the following:
+			// - If keepalive dormancy conditions are met, then go dormant.
+			// - If there is an active pinger routine, then do nothing.
+			// - If dormany conditions are not met and there is no active pinger
+			//   routine, start one.
+			sendPing := true
+			goDormant := func() bool {
+				t.mu.Lock()
+				defer t.mu.Unlock()
+				if len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream {
+					// If pinger routine is active, it probably means that one was started
+					// when there was an active stream, but now there is none, and
+					// therefore we should stop the pinger.
+					if done != nil {
+						close(done)
+						done = nil
+					}
+					// Make awakenKeepalive writable.
+					<-t.awakenKeepalive
+					return true
+				}
+				return false
+			}()
+			if goDormant {
 				select {
 				case <-t.awakenKeepalive:
-					// If the control gets here a ping has been sent
-					// need to reset the timer with keepalive.Timeout.
+					// If the control gets here, it means that a ping was sent right
+					// after stream creation. We need to make sure we get an ack for it
+					// before kp.Timeout expires.
+					sendPing = false
```

Comment:
I didn't understand how this can happen. Can you explain this?

Reply:
In `http2Client.NewStream()`, in the case where the number of active streams has changed from 0 to 1, we write on the `awakenKeepalive` channel to awaken the `keepalive` goroutine. After sending on the `awakenKeepalive` channel, we also send out a PING. This implementation detail is based on the proposal A8. Related excerpt here:
`
Since keepalive is not occurring on HTTP/2 connections without any streams, there will be a higher chance of failure for new RPCs following a long period of inactivity. To reduce the tail latency for these RPCs, it is important to not reset the keepalive time when a connection becomes active; if a new stream is created and there has been greater than 'keepalive time' since the last read byte, then a keepalive PING should be sent (ideally before the HEADERS frame). Doing so detects the broken connection with a latency of keepalive timeout instead of keepalive time + timeout.
`

So, in our `keepalive` routine, when we are awakened we know that a PING has already been sent. So, we start the `pinger` and ask it to not send out a new PING.